feat(docs): added accessibility guide

karsa-mistmere opened the pull request:
closes #2121 

## What is the purpose of this pull request?
- [x] Documentation update

## Before Submitting <!-- For every PR! -->
<!-- All of these requirements must be fulfilled. -->
- [x] I've read the [Contribution Guidelines](https://github.com/lucide-icons/lucide/blob/main/CONTRIBUTING.md).
- [x] I've checked if there was an existing PR that solves the same issue.

jguddas commented on docs/images/a11y/alttext-buttons.svg:
Would be nice to have a better text here other than just `Button`.

jguddas commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,114 @@
+
+![](../../images/a11y/alttext-buttons.svg?raw=true)
+
+### On icon buttons
+
+Icon buttons are buttons that do not contain any visible text apart from the icon itself (think of
+the close button of a dialog for example).
+
+![](../../images/a11y/alttext-iconbuttons.svg?raw=true)
```
Why?

jguddas reviewed (changes requested):
- The content of the images is not accessible, the alt text is different from the content of the image. I would generally prefer if we moved the text out of the images and made them just normal text.
- Would be nice to have a section on flashing images and use reduce motion as well
- Also `aria-label` is not perfect, using something like RadixUIs `AccessibleIcon` is something that we should point people to use instead.
- A section on who this for and why you should do this would also be nice.

jguddas commented:
We can also point people to resources that they should reference, be it just the WCAG or even component libraries or guides like https://inclusive-components.design/toggle-button/

karsa-mistmere commented:
> * The content of the images is not accessible, the alt text is different from the content of the image. I would generally prefer if we moved the text out of the images and made them just normal text.

For the most part this is entirely intentional, these are illustrations, I've added accessible text where there was information otherwise missing, but most of these illustrations contain redundant information that is not useful to be read out as is. There's probably room for improvement, on that I can agree.

> Would be nice to have a section on flashing images and use reduce motion as well

We don't really have ootb support for such animations, so I'm not entirely sure if it's relevant within the context of Lucide.

> Also aria-label is not perfect, using something like RadixUIs AccessibleIcon is something that we should point people to use instead.

Well, the current guide does recommend not using it in places where it's problematic, but you're right, there probably space for some more nuance about this.

> A section on who this for and why you should do this would also be nice.

Is it not for everyone using Lucide? 🧌

jguddas commented:
> > A section on who this for and why you should do this would also be nice.
> 
> Is it not for everyone using Lucide? 🧌

Example: Your should put enough contrast so people with visual impairments, be it physical or just them being outside in the glaring sun can still get the amazing benefit of looking at those cute icons.

jguddas commented:
> > Would be nice to have a section on flashing images and use reduce motion as well
> 
> We don't really have ootb support for such animations, so I'm not entirely sure if it's relevant within the context of Lucide.

Tailwind based example: You can do `<DogIcon className="animate-ping" />` but what you should consider using `<DogIcon className="motion-safe:animate-ping" />` instead.

karsa-mistmere commented:
> > > Would be nice to have a section on flashing images and use reduce motion as well
> > 
> > 
> > We don't really have ootb support for such animations, so I'm not entirely sure if it's relevant within the context of Lucide.
> 
> Tailwind based example: You can do `<DogIcon className="animate-ping" />` but what you should consider using `<DogIcon className="motion-safe:animate-ping" />` instead.

Yeah, but this is entirely dependant on Tailwind, I think it should be their responsibility to educate people about using their animations responsibly. :)

(To be frank, this should be their default, it shouldn't be something to opt into.)

jguddas commented on docs/guide/advanced/accessibility.md:
```diff
@@ -117,6 +116,36 @@ contained icon.
 
+## A note on `aria-label`
+
+Although you could provide accessible labels to your elements via the `aria-label` attribute, we
+generally suggest using your chosen CSS framework's "visually hidden" utility instead whenever
+possible, since some assistive technology fails to announce `aria-label` attributes in case the
+element is not of an appropriate semantic role.
```
The main reason I don't like `aria-label` is that it gets exposed by all major browsers as AXDescription while normal content is exposed as AXTitle, and according to the spec, AXDescription should be read out in the system language, meaning your `lang=…` attribute gets ignored and your screen reader might sound funky.

karsa-mistmere replied:
Neato, I've updated this description and removed all traces of `aria-label` from the illustrations.

jguddas commented on docs/images/a11y/alttext-buttons.svg:
We could even set `aria-hidden="true"` on decorative icons like that.

karsa-mistmere replied:
I'm a bit confused, does this mean that simple SVGs with no accessible text are currently being read out? 🤔

jguddas replied:
The read-out text for voiceover for example is "You are currently on an image. To begin interacting with the contents of this image, press Control-Option-Shift-Down Arrow.", that's so super not helpful for decorative images.


<img width="729" alt="image" src="https://github.com/lucide-icons/lucide/assets/25524993/da3a605c-c088-4b71-89bb-fd7299f65f23">

karsa-mistmere replied:
Oh jeez louise, that's awful. 🤮

+1 for adding aria-hidden then. Best course of action would actually be to automatically add that no matter what.

jguddas replied:
Hiding the icon by default sounds super dangerous.

As I see it there are 2 okayish options:

1. Add a default label that is a descriptive representation of the icon. #2121
2. Make setting a `label` / `aria-hidden` attribute mandatory and well documented for when to use what.

karsa-mistmere replied:
> 1. Add a default label that is a descriptive representation of the icon. https://github.com/lucide-icons/lucide/discussions/2121

Terrible-terrible idea, as I've already explained in #2121:
- Accessible labels on functional icons should describe the use case, not the graphical representation (and even that isn't necessarily unambiguous, especially considering abstract icons).
- Decorative icons on the other hand should have no accessible label whatsoever.

> 2. Make setting a label / aria-hidden attribute mandatory and well documented for when to use what.

Supporting any kind of accessible label would lead to breaking changes, since we cannot use either `<title>` nor `aria-label="..."` as already discussed in previous threads. That means we would have to add at least one extra element to the DOM and inject CSS where there previously were none (or use inline CSS for our `visually-hidden` util).

> Hiding the icon by default sounds super dangerous.

I don't really see why, it is practically _never_ advisable to set an accessible label on an icon itself, but instead on other interactive elements, or using some kind of `visually-hidden`/`sr-only` util class, but that by nature will be on a different element as well.

ericfennis commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,158 @@
+## A note on `aria-label`
+
+Although you could provide accessible labels to your elements via the `aria-label` attribute, we
+generally recommend against this for and instead suggest that
+you [use your chosen CSS framework's "visually hidden" utility whenever possible](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
```
Can you explain why we should be "against" this? It would be nice if we explained this more in the text, about why using "visually hidden" is better than using `aria-label`, we also could provide some sources for blog posts or documentation to prove our "recommendation".

Also maybe this reads a bit nicer
``` suggestion
generally recommend visually hiding textual descriptions. For example by [using utility CSS framework's "visually hidden"](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
```

karsa-mistmere replied:
That means something different though. Also, the link itself is the resource that explains why `.visually-hidden` > `[aria-label]`, I've rephrased the link text so that it's clearer.

```suggestion
generally recommend against this and instead suggest that you use your chosen CSS framework's "visually hidden" utility whenever possible. You can [read more about why `aria-label` might not be the best solution](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
```

ericfennis replied:
I'm still not happy with the phraise "we generally recommend against this". That suggest that we are against the use of `aria-label`. And from what I found there nothing wrong with it.
I think you should approach it more positively. "If you are using a CSS framework like Tailwind, we recommend using visually hidden utilities whenever possible"

jguddas replied:
The options from best to worst:

1. visible label
2. visually hidden label
3. aria-label

karsa-mistmere replied:
> And from what I found there nothing wrong with it.

Well, there kind of is, that's what the article linked is about: https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/#the-aria-label-attribute-is-not-always-translated

But sure, we can rephrase it to be more lenient.

ericfennis replied:
They use the word "avoid" instead of "against", which is I think more right fit. Since using `aria-label` is not wrong. I  think we also should link this article to this paragraph, it gives a bit more context to this topic.

karsa-mistmere replied:
No, in certain scenarios it might actually be wrong, since `aria-label` will generally be ignored on non-interactive elements. I don't believe linking to a certain paragraph is necessary, since the entirety of article is about why `aria-label` should be avoided if possible, there isn't a single reason, there are multiple.

Since this thread is on an outdated commit, I've made a suggestion here: https://github.com/lucide-icons/lucide/pull/2122/files#r1636386974

ericfennis replied:
Ah ok true, tnx for suggestion think it reads a bit nicer as well.

ericfennis reviewed (commented):
Sorry for the late review.
This is such an awesome guide, thank you @karsa-mistmere super useful!
We can improve this even more if we add some small code examples for some things to make it visible in the code. 

I've provided some suggestions and one comment

karsa-mistmere commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,158 @@
+screen readers, leading to nonsensical text.
+
+![](../../images/a11y/alttext-buttons.svg?raw=true)
+
```
@ericfennis: These could be useful, but they break up the guide quite a bit, I've hidden them within details elements.

ericfennis replied:
Nice looks great!

karsa-mistmere commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,204 @@
+## A note on `aria-label`
+
+Although you could provide accessible labels to your elements via the `aria-label` attribute, we
+generally recommend against this and instead suggest that you use your chosen CSS framework's "
```
```suggestion
generally recommend avoiding this and instead suggest that you use your chosen CSS framework's "
```

jguddas commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,204 @@
+visually hidden" utility whenever possible. You can
+[read more about why `aria-label` might not be the best solution](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
+
+E.g. in case of Bootstrap:
```
```suggestion
### Example - Bootstrap
```

jguddas commented on docs/guide/advanced/accessibility.md:
````diff
@@ -1,4 +1,204 @@
+</div>
+```
+
+Or with tailwindcss:
````
```suggestion
### Example - Tailwind CSS
```

jguddas commented on docs/guide/advanced/accessibility.md:
```diff
@@ -1,4 +1,204 @@
+generally recommend avoiding this and instead suggest that you use your chosen CSS framework's "
+visually hidden" utility whenever possible. You can
+[read more about why `aria-label` might not be the best solution](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
+
```
```suggestion

### Example - Radix UI

https://www.radix-ui.com/primitives/docs/utilities/accessible-icon

***tsx
import { ArrowRightIcon } from 'lucide-react';
import { AccessibleIcon } from '@radix-ui/react-accessible-icon';

<AccessibleIcon label="Next item">
  <ArrowRightIcon />
</AccessibleIcon>
***
```

karsa-mistmere replied:
https://www.radix-ui.com/primitives/docs/utilities/accessible-icon

Ought to be a link with descriptive text.
````suggestion
### Example - Radix UI

Use [Radix UI's built-in accessible icon utility component](https://www.radix-ui.com/primitives/docs/utilities/accessible-icon).

```tsx
import { ArrowRightIcon } from 'lucide-react';
import { AccessibleIcon } from '@radix-ui/react-accessible-icon';

<AccessibleIcon label="Next item">
  <ArrowRightIcon />
</AccessibleIcon>
```
````
